src/analyses/variable-sensitivity/array_abstract_object.cpp

```diff
@@ -35,6 +35,21 @@ array_abstract_objectt::array_abstract_objectt(
   PRECONDITION(e.type().id() == ID_array);
 }
 
+abstract_object_pointert array_abstract_objectt::expression_transform(
+  const exprt &expr,
+  const std::vector<abstract_object_pointert> &operands,
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  if(expr.id() == ID_index)
+  {
+    return read(environment, expr, ns);
```

Comment:
Given that `read` just calls `read_index` can you call `read_index` directly here (and the corresponding functions in struct, pointer, etc.) and then remove `read` from the interface of `abstract_objectt` completely.